File: selfdrive/ui/qt/setup/reset.cc
#include <QApplication>
#include <QHBoxLayout>
#include <QLabel>
#include <QPushButton>
#include <QTimer>
#include <QVBoxLayout>

#include "selfdrive/ui/qt/qt_window.h"
#include "selfdrive/ui/qt/setup/reset.h"

#define NVME "/dev/nvme0n1"
#define USERDATA "/dev/disk/by-partlabel/userdata"

void Reset::doErase() {
  // best effort to wipe nvme
  std::system("sudo umount " NVME);
  std::system("yes | sudo mkfs.ext4 " NVME);

  int rm = std::system("sudo rm -rf /data/*");
  std::system("sudo umount " USERDATA);
  int fmt = std::system("yes | sudo mkfs.ext4 " USERDATA);

  if (rm == 0 || fmt == 0) {
    std::system("sudo reboot");
  }
  body->setText(tr("Reset failed. Reboot to try again."));
  rebootBtn->show();
}

void Reset::startReset() {
  body->setText(tr("Resetting device...\nThis may take up to a minute."));
  rejectBtn->hide();
  rebootBtn->hide();
  confirmBtn->hide();
#ifdef __aarch64__
  QTimer::singleShot(100, this, &Reset::doErase);
#endif
}

void Reset::confirm() {
  const QString confirm_txt = tr("Are you sure you want to reset your device?");
  if (body->text() != confirm_txt) {
    body->setText(confirm_txt);
  } else {
    startReset();
  }
}

Reset::Reset(ResetMode mode, QWidget *parent) : QWidget(parent) {
  QVBoxLayout *main_layout = new QVBoxLayout(this);
  main_layout->setContentsMargins(45, 220, 45, 45);
  main_layout->setSpacing(0);

  QLabel *title = new QLabel(tr("System Reset"));
  title->setStyleSheet("font-size: 90px; font-weight: 600;");
  main_layout->addWidget(title, 0, Qt::AlignTop | Qt::AlignLeft);

  main_layout->addSpacing(60);

  body = new QLabel(tr("System reset triggered. Press confirm to erase all content and settings. Press cancel to resume boot."));
  body->setWordWrap(true);
  body->setStyleSheet("font-size: 80px; font-weight: light;");
  main_layout->addWidget(body, 1, Qt::AlignTop | Qt::AlignLeft);

  QHBoxLayout *blayout = new QHBoxLayout();
  main_layout->addLayout(blayout);
  blayout->setSpacing(50);

  rejectBtn = new QPushButton(tr("Cancel"));
  blayout->addWidget(rejectBtn);
  QObject::connect(rejectBtn, &QPushButton::clicked, QCoreApplication::instance(), &QCoreApplication::quit);

  rebootBtn = new QPushButton(tr("Reboot"));
  blayout->addWidget(rebootBtn);
#ifdef __aarch64__
  QObject::connect(rebootBtn, &QPushButton::clicked, [=]{
    std::system("sudo reboot");
  });
#endif

  confirmBtn = new QPushButton(tr("Confirm"));
  confirmBtn->setStyleSheet(R"(
    QPushButton {
      background-color: #465BEA;
    }
    QPushButton:pressed {
      background-color: #3049F4;
    }
  )");
  blayout->addWidget(confirmBtn);
  QObject::connect(confirmBtn, &QPushButton::clicked, this, &Reset::confirm);

  bool recover = mode == ResetMode::RECOVER;
  rejectBtn->setVisible(!recover);
  rebootBtn->setVisible(recover);
  if (recover) {
    body->setText(tr("Unable to mount data partition. Partition may be corrupted. Press confirm to erase and reset your device."));
  }

  setStyleSheet(R"(
    * {
      font-family: Inter;
      color: white;
      background-color: black;
    }
    QLabel {
      margin-left: 140;
    }
    QPushButton {
      height: 160;
      font-size: 55px;
      font-weight: 400;
      border-radius: 10px;
      background-color: #333333;
    }
    QPushButton:pressed {
      background-color: #444444;
    }
  )");
}

int main(int argc, char *argv[]) {
  ResetMode mode = ResetMode::USER_RESET;
  if (argc > 1) {
    if (strcmp(argv[1], "--recover") == 0) {
      mode = ResetMode::RECOVER;
    }
  }

  QApplication a(argc, argv);
  Reset reset(mode);
  setMainWindow(&reset);
  return a.exec();
}

File: system/ui/raylib/reset.py
import pyray as rl
import os
import sys
import time
import threading
from enum import IntEnum
from openpilot.system.ui.raylib.gui.button import gui_button
from openpilot.system.ui.raylib.gui.label import gui_label
from openpilot.system.ui.raylib.gui.application import gui_app


NVME = "/dev/nvme0n1"
USERDATA = "/dev/disk/by-partlabel/userdata"

def do_erase():
  # Best effort to wipe NVME
  # os.system(f"sudo umount {NVME}")
  # os.system(f"yes | sudo mkfs.ext4 {NVME}")

  # # Removing data and formatting
  # rm = os.system("sudo rm -rf /data/*")
  # os.system(f"sudo umount {USERDATA}")
  # fmt = os.system(f"yes | sudo mkfs.ext4 {USERDATA}")

  # if rm == 0 or fmt == 0:
  #   os.system("sudo reboot")
  return "Reset failed. Reboot to try again."

class ResetMode(IntEnum):
  USER_RESET = 0
  RECOVER = 1
  FORMAT = 2

class Reset:
  def __init__(self, mode):
    self.mode = mode
    self.confirm_msg = "Are you sure you want to reset your device?"
    self.body_text = "System reset triggered. Press confirm to erase all content and settings. Press cancel to resume boot."
    self.resetting = False

  def start_reset(self):
    self.body_text = "Resetting device...\nThis may take up to a minute."
    self.resetting = True
    threading.Timer(0.1, do_erase).start()

  def confirm(self):
    if self.body_text != self.confirm_msg:
      self.body_text = self.confirm_msg
    else:
      self.start_reset()

def draw_reset_ui(reset):
    rl.begin_drawing()
    rl.clear_background(rl.BLACK)
    left_margin, top_margin, right_margin, bottom_margin = 45, 220, 45, 45
    content_rect = rl.Rectangle(left_margin, top_margin,
                                rl.get_screen_width() - (left_margin + right_margin),
                                rl.get_screen_height() - (top_margin + bottom_margin))
    # Title
    label_rect = content_rect
    label_rect.x += 140
    label_rect.width -= 280
    gui_label(label_rect, "System Reset", 90)

    # Body text
    label_rect.y += 90 + 60
    gui_label(label_rect, reset.body_text, 80)

    # Buttons
    button_height = 160
    button_top = rl.get_screen_height() - button_height - bottom_margin
    button_width = (rl.get_screen_width() - (left_margin + right_margin + 50)) / 2.0
    rl.gui_set_style(rl.GuiControl.DEFAULT, rl.GuiDefaultProperty.TEXT_SIZE, 55)
    if not reset.resetting:
      if gui_button(rl.Rectangle(left_margin, button_top, button_width, button_height), "Cancel"):
        rl.close_window()
      if reset.resetting:
        if gui_button(rl.Rectangle(left_margin + button_width + 50, button_top, button_width, button_height), "Reboot"):
          os.system("sudo reboot")
      else:
        if gui_button(rl.Rectangle(left_margin + button_width + 50, button_top, button_width, button_height), "Confirm", rl.Color(70, 91, 234, 255)):
          reset.confirm()

    rl.end_drawing()


def main():
  gui_app.init_window("Reset", 20)

  mode = ResetMode.USER_RESET
  if len(sys.argv) > 1:
    if sys.argv[1] == "--recover":
      mode = ResetMode.RECOVER
    elif sys.argv[1] == "--format":
      mode = ResetMode.FORMAT

  reset = Reset(mode)

  while not rl.window_should_close():
    draw_reset_ui(reset)

  gui_app.close()


if __name__ == "__main__":
  main()
